[1단계 - 지하철 정보 관리 기능] 도기(김동호) 미션 제출합니다.

kdkdhoho opened the pull request:
안녕하세요 검프! 도기입니다 🐶
이번 미션은 구현에 실패했습니다.
대신 ~반성문 쓰듯이~ 페어와 의견이 갈린 부분에 대해 함께 정리하고 각자 리뷰어에게 물어보기로 결정했습니다.

구현은 실패했지만, 미션을 진행하며 페어와 정말 많은 이야기를 나누었고 그 과정들이 너무 값졌다고 생각합니다.
오히려 구현을 모두 끝냈을 때보다 더 많은 것이 남은 기분이에요.
그래도 아직 미션이 끝나지 않았으니, 끝날 때까지 최선을 다해보겠습니다.

---

# 페어와의 의견 정리

## 1. DB 설계가 먼저냐? or 도메인 설계가 먼저냐?

### DB 설계가 먼저라고 생각한 이유 (페어)
- 변경 가능성이 큰 도메인 보다 변경 가능성이 적고 어려운 테이블 설계를 먼저 해놓는 게 낫다라고 생각했습니다.
- DB 설계를 먼저 해놓으면 도메인 설계는 해당 테이블에 맞춰서 하면 되겠다고 생각했습니다.

### 도메인 설계가 먼저라고 생각한 이유 (도기)
-  변경이 잦은 View, DB와는 달리 프로그램의 뿌리와 같은 도메인을 먼저 개발하고 싶었다.
- 프로그램의 유지보수를 위해 도메인부터 견고히 쌓아 올리고 싶었다.
- DB를 먼저 설계하면, 데이터를 위한 서버가 된다고 생각했다.
- 복잡한 요구사항을 보고, 어떻게 테이블을 설계할 지 도통 감이 잡히지 않았다. 그렇기에 우리가 그나마 할 수 있는 가장 작은 단위의 도메인부터 개발하고 싶었다.
- 배포 후에는 테이블을 수정하는 것은 매우 큰 비용이 든다고 생각한다. 하지만 개발 단계에서는 테이블을 수정하는 것은 쉽다고 생각한다.

만약 검프라면 어떻게 설계하셨을건가요?

---

## 2. 좋은 URI 설계를 어떻게 해야 할까?

노선에 역을 추가하는 API를 설계할 때 2개의 역을 추가하는 케이스, 1개의 역을 추가하는 케이스로 나뉩니다. 이때 필요한 요청 값이 다르고, URI path를 어떻게 정해야 이해하기 쉬울지 많은 고민을 했습니다.

### 1번 째 - 역 두 개: "/lines/{lineId}/stations" 역 하나: "/lines/{lineId}/stations/{stationId}"
노선에 역 제거 URI와 역 등록하는 URI를 통일하고 싶어서 이렇게 설계를 했다가 역 하나를 등록할 때 아직 등록되지 않은 {stationId}를 URI에 포함하는게 이상하다고 느껴 수정하게 되었습니다.

### 2번 째 - 역 두 개, 하나(통일): "/lines/{lineId}?staitonId= "
쿼리 파라미터 값의 유무에 따라 두 개를 추가할 지 하나를 추가할 지 어플리케이션 상에서 결정하도록 설계 했습니다. 하지만 쿼리파라미터는 검색, 필터링, 페이징 등에만 사용한다고 알게 되어 수정했습니다.

### 3번 째 - 역 두 개: "/lines/{lineId}" 역 하나: "/lines/{lineId}/stations"
들어오는 요청에 null을 포함하기 보다 조건에 따라 URI를 나눈게 좋다고 생각하였습니다. 노선에 연결되는 리소스가 Station 하나라서 초기 역 추가 시에는 “/lines/{lineId}" 를 사용하였고 이 후에는 Station이 추가된 후에는 "/lines/{lineId}/stations" 를 사용하였습니다.
하지만 lines에 station이 아닌 다른 리소스가 추가된다면 /lines/{lineId}" URI를 사용하지 못할거 같고, 역 하나를 추가하는 URI와 통일성이 없다고 생각합니다.

검프라면 어떻게 하셨을 지 궁금합니다.

---

## 3. 외부 라이브러리(Graph)를 사용하는게 맞나?
### 고민을 하게된 이유
작은 단위의 도메인부터 구현을 진행하다가 노선에 역을 추가하는 기능을 구현하려다가 고민이 시작되었습니다. 구현한 도메인에는 거리 개념이 포함되지 않았고, 노선에 역을 추가하는 기능을 구현하기 위해서는 거리 개념이 필요해졌습니다. 따라서 그래프 자료구조를 적용한 객체를 설계하자는 의견이 나왔고, 해당 객체를 구현할 때 어떤 방법으로 구현할 지에 대해 의견이 나뉘었습니다.

### Graph 라이브러리를 사용하면 안된다고 생각한 이유 (페어)
- 외부 라이브러리를 의존하지 않고도 지하철 도메인에 필요한 그래프 자료구조를 구현할 수 있다고 생각했습니다. 그리고 다음 단계에서 다익스트라 알고리즘을 사용할 때 해당 라이브러리를 추가해도 충분하다고 생각했습니다.

### Graph 라이브러리를 사용해도 된다고 생각한 이유 (도기)
- 자료구조를 직접 고민하고 구현하는 시간에 복잡한 요구사항을 고민하는 시간을 더 가지고 싶었습니다.
- 시간이 충분했다면 외부 라이브러리에 의존하지 않았겠지만, 짧은 시간 내에 문제를 해결해야하는 상황에서는 적극적으로 사용하여 시간을 단축하고 싶었습니다.

검프는 어떻게 생각하시나요?

---

## 4. 도메인과 DB 불일치를 해소 시킬 때
### 고민을 하게된 이유
노선에 역을 추가하는 기능을 구현하다보니, 도메인과 테이블의 불일치가 존재했습니다. 이 둘을 변환해서 사용하자는 것에는 이견이 없었지만 변환하는 방법에 대해 의견이 나뉘었습니다.

### 도메인 객체를 하나로 합치는 객체에서 해소하자 (페어)
- 변환하는 과정에서 SECTION 테이블에서 연결된 역 정보와 호선 정보를 모두 불러와서 조립할 때 도메인 객체에서 호선을 관리하는 객체와 역을 관리하는 객체가 다르기 때문에 하나의 객체로 묶어서 변환하는게 DB에서 불러올때 사용되는 리소스나 역을 추가할 때 같은 객체에서 두 정보를 가지고 있으면 관리하기가 편하다고 판단했습니다.
그 후 분리의 필요성이 느껴지면 분리해도 충분하다고 판단했습니다.

### 도메인 객체 별로 해소하자 (도기)
- 기존에 존재하는 도메인은 그대로 둔 채, Repository 계층에서 각 도메인 객체를 조립해주는 역할만 추가하면 된다고 생각했습니다. 그리고 Domain에서 복잡한 비즈니스 로직을 수행하고 그에 대한 결과를 Controller로 반환하거나 Repository를 통해 저장하는 구조를 택한다면, 잘 나뉘어진 추상화 계층이라고도 생각했습니다. 또한, 도메인에 DB와 관련된 정보(id)를 포함한다면 도메인이 DB를 의존하게 되므로 더욱 도메인을 DB로부터 보호하고 싶었습니다.

검프라면 어떻게 하셨나요?

---

## 5. 실무라면?
지금 저희는 교육 단계이기에 기간 내에 완성을 하지 못해도, 학습하는 것이 우선이라고 생각합니다. 하지만 실무에서는 학습보다는 결과물이 우선입니다.
만약 저희가 실무에서 같은 상황(시간이 부족하고 서로 설득이 안되는 상황)이었다면 규모가 있는 기업의 경우 시니어 개발자의 의견을 물어보고 결정했을 것 같습니다. 하지만 주변에 도움을 요청할 사람이 없는 상황의 스타트업의 경우, 지인이나 커뮤니티에 의견을 물어보고 더 많은 의견으로 결정했을 것 같습니다.

만약 검프라면 어떻게 해결하셨을까요?

This2sho committed:
docs: 요구사항 분석 정리

This2sho committed:
feat(Distance): 거리 VO 구현

This2sho committed:
feat(Stations): 역을 추가하는 기능 구현

This2sho committed:
feat(Line): 노선에 역을 추가하는 기능 구현

This2sho committed:
build: graph 라이브러리 의존성 추가

This2sho committed:
test(LineIntegrationTest): 라인 통합 테스트 작성

This2sho committed:
refactor(Station): 생성자 수정

This2sho committed:
feat(StationAddRequest): 노선에 역을 추가하는 요청 객체 구현

This2sho committed:
feat(SubwayMap): 노선도에서 역을 추가, 삭제하는 기능 구현

This2sho committed:
build: DB 설정 추가

This2sho committed:
refactor(Direction): Direction 삭제

This2sho committed:
docs(README): TABLE 설계 수정

This2sho committed:
feat(Section): Section 테이블 설계 및 insert 기능 구현

This2sho committed:
feat(SubwayMap): 전체 노선도에 Station을 추가하는 기능 구현

This2sho committed:
refactor(StationAddRequest): 변수명 변경

This2sho committed:
feat: 빈 노선에 역을 추가하는 기능 구현

This2sho committed:
feat(Lines): 노선들을 모아놓은 객체 생성

This2sho committed:
feat(GlobalExceptionHandler): 존재하지 않는 id가 들어왔을 때 처리하는 기능 구현

kdkdhoho committed:
docs(README): 요구사항 재정리

kdkdhoho committed:
chore: 프로덕션 DB 설정에서 테스트 DB 설정으로 수정

kdkdhoho committed:
docs(README): 노선에 역 1개 추가 시, 요청 데이터 수정

kdkdhoho committed:
docs(README): 기능 요구사항 세분화

kdkdhoho committed:
feat(StationAlreadyExistException): 역 추가 시, 기존에 동일한 역이 존재하는 커스텀 예외 추가

kdkdhoho committed:
feat(IllegalStationAddException): 역을 추가할 때 비정상적인 요청에 대한 커스텀 예외 추가

kdkdhoho committed:
refactor(Stations): 역을 추가하는 기능 상세화

kdkdhoho committed:
refactor(Line): 내부에 Stations를 가지지 않도록 수정

kdkdhoho committed:
refactor(Stations): 생성자, getter 타입을 List로 수정

kdkdhoho committed:
rename: 역을 추가할 때 발생하는 커스텀 예외 네이밍 수정

kdkdhoho committed:
feat(LineNotFoundException): 존재하지 않는 역 조회 시 커스텀 예외 추가

Livenow14 commented on README.md:
````diff
@@ -1 +1,188 @@
+
+## 노선 목록 조회
+
+```http request
+GET "/lines"
+
+// List<LineResponse>
+HTTP Status : 200 OK
+[
+    {
+        "id": ?,
+        "name": ?,
+        "color": ?
+    },
+    {
+        ...
+    }
+]
+```
+
+## 노선에 초기 역 추가
+
+```http request
+POST "/lines/{lineId}"
+
+// StationAddRequest (A - B)
+{
+    "previousStationId" : A,
+    "nextStationId" : B,
+    "distance" : 11
+}
+```
````
Intellij에서 지원하는 https://www.jetbrains.com/help/idea/http-client-in-product-code-editor.html
를 사용해보시는건 어떨까요?

kdkdhoho replied:
이런 꿀기능이 있었군요. 한번 적용해보겠습니다. 감사합니다 😀

Livenow14 commented on README.md:
````diff
@@ -1 +1,188 @@
+
+# DB TABLE 설계
+
+```mysql
+CREATE TABLE STATION
+(
+    id   BIGINT NOT NULL AUTO_INCREMENT,
+    name VARCHAR(255) NOT NULL UNIQUE,
+    PRIMARY KEY (id)
+);
+```
+
+```mysql
+CREATE TABLE LINE
+(
+    id    BIGINT AUTO_INCREMENT NOT NULL,
+    name  VARCHAR(255)          NOT NULL UNIQUE,
+    color VARCHAR(255)          NOT NULL UNIQUE,
+    PRIMARY KEY (id)
+);
+```
+
+```mysql
+CREATE TABLE SECTION
+(
+    id         BIGINT AUTO_INCREMENT NOT NULL,
+    line_id    BIGINT,
+    previous_station_id BIGINT,
+    next_station_id BIGINT,
+    distance INT NOT NULL,
+    PRIMARY KEY (id),
+    FOREIGN KEY (line_id) REFERENCES LINE (id) ON DELETE CASCADE,
+    FOREIGN KEY (previous_station_id) REFERENCES STATION (id) ON DELETE CASCADE,
+    FOREIGN KEY (next_station_id) REFERENCES STATION (id) ON DELETE CASCADE
+);
+```
````
DB schema파일을 작성 후 docs에 링크를 달 수 있어요

Livenow14 commented on src/main/java/subway/domain/Line.java:
```diff
@@ -3,34 +3,58 @@
+    private final Long id;
+    private final String name;
+    private final String color;
+    private final Stations stations;
```
Line이 Station을 가지게 된다면, 객체 모델은 어떻게 그려질까요? 

Line이 Station을 알고 있는것 보다, Section이 Line을 알고있는게 조금더 유연한 설계라 생각돼서요. 

(매번 섹션이 바뀔때 마다, 라인의 수정이 이뤄져야함)

kdkdhoho replied:
`Lines -> Line -> Stations -> Station`의 객체 모델이 될 것 같아요.
이렇게 되면 모든 호선을 Lines가 관리하고, 각 호선에 포함되는 역들을 Line을 통해 관리할 수 있을 것 같아요.

Section에 대한 모델은 SubwayMap을 통해 적용하려고 했습니다!

kdkdhoho replied:
아래 피드백과 함께 고려 중인 사항이 있습니다.
Graph 라이브러리를 사용하지 않고, `Section`을 도메인 객체로서 사용하는 것으로 생각 중인데요.
검프께서 Section이 Line을 알고 있는 구조가 더 나은 설계라고 해주셨어요. 그런데 Section들이 모인 것이 결국 하나의 Line이라고 생각됩니다. 즉, `Line`이 `List<Section>`을 알고 있는 구조가 될 것 같아요.

이에 대해 검프의 의견을 듣고 싶습니다!

Livenow14 replied:
이건 모델링의 성격에 따라 달라질 것 같아서요 :) 
저는 일대다 관계보다 다대일 관계를 선호한느 편입니다 :)

kdkdhoho replied:
아래와 같이 이해했습니다.
```
라인 입장에서, 여러 섹션을 갖는 것 - 일대다 관계
섹션이 여러 라인 중 하나를 골라 갖는 것 - 다대일 관계
```
제가 정확히 이해했을까요?

그리고 다대일 관계를 선호하는 이유가 무엇인지 궁금합니다!

Livenow14 replied:
일대다 관계로 작성되다보면, 변경에 있어서 조금 불편해지는 경향이 있어서요. 

일대다 관계시 결국 양쪽 모두 서로를 의존하기에 수정, 추가시 양쪽 관계를 다 고려해야하는 문제가 생기는데 (db와 객체의 임피던스 불일치). 초반 개발 진행시 고려사항이 많아지는거라 제한하고 있어요. 

물론 DDD관점에서는 일대다를 선호하긴 합니다.

kdkdhoho replied:
일대다, 다대일. 하나의 기준을 잡고 나눠서 생각해봐야 할 것 같아요.
그렇다면 그 기준은 어떻게 되나요??

Livenow14 replied:
https://github.com/woowacourse/jwp-subway-path/pull/84#discussion_r1198727595 

구현하려는 개발자의 모델링 성격에 따라 달라질 것 같은데요. 
기준은 과감히 없다 생각해주시면 좋을 것 같네요. 

도기가 생각하시기에 효율적인 방법을 고려해보셨으면 좋겠어요. 
제가 제안 드린 부분은, 1단계 구현이 다 완료되지 않았는데 일대다와 같이 변경 고려사항이 많은 모델링까지 필요했을까 라는 의도에서 남겼습니다.

Livenow14 commented on src/main/java/subway/domain/SubwayMap.java:
```diff
@@ -0,0 +1,57 @@
+    }
+
+    public void addStation(Station upStation, Station downStation, Integer distance) {
+        subwayMap.addVertex(upStation);
+        subwayMap.addVertex(downStation);
+        DefaultWeightedEdge edge = subwayMap.addEdge(upStation, downStation);
+        subwayMap.setEdgeWeight(edge, distance);
```
현재 요구사항이 그렇게 복잡하지 않은데, 해당 기술(SimpleDirectedWeightedGraph)을 사용해야할 이유가 있을까요?!?! 

조금 과한 기술 선택으로 보여서요 :)

kdkdhoho replied:
도메인 객체를 처음에 `Lines -> Line -> Stations -> Station` 의 구조로 설계했습니다.
그리고 노선에 역을 추가하는 기능을 구현하려고 하니 거리와 방향 개념이 필요하게 됐습니다.
이때 자연스럽게 그래프 자료구조를 적용해야겠다는 판단이 들었고, 존재하는 라이브러리를 사용하여 역과 역 사이의 거리(가중치)와 방향을 적용하기 위해 사용했습니다.

그렇다면 검프는 라이브러리를 사용하지 않고 직접 그래프 자료구조를 구현하실 생각이신가요?

Livenow14 replied:
현재 요구사항에 필요하지 않음에도 자연스럽게 라이브러리가 떠오르신걸까요?

kdkdhoho replied:
1단계 미션 소개 페이지에 객체 설계 힌트로 그래프 자료구조에 대한 소개와 함께 라이브러리에 대한 정보도 함께 제공이 되었습니다.

물론 1단계에서 라이브러리를 써야 하는 이유는 언급되지 않았으나, 미션을 진행하며 자연스레 그래프 자료구조를 적용해야겠다고 생각이 들었고 그에 따라 라이브러리가 떠올랐습니다.

사용하고 싶었던 가장 큰 이유는, 제한된 시간 내에 문제를 해결해야 했기에 직접 자료구조를 구현하는 시간을 절약하고 싶었기 때문입니다.

Livenow14 replied:
질문의 의도는 "현재 요구사항에 라이브러리가 필요하나?" 입니다. 

현재 요구사항에서 가중치 계산같은 복잡한 구현은 필요하지 않아 보여서요.

kdkdhoho replied:
현재 요구사항에 라이브러리가 필요하지 않습니다.
하지만, 역을 중간에 추가, 삭제하는 기능을 구현할 때 거리 계산이 필요했고 이때 거리의 개념을 "가중치"로 생각하면 쉽게 구현할 수 있을 것 같았어요..!
물론 다른 방식으로 해결할 수 있겠지만, 당시엔 마감까지 시간이 너무 없었고 다른 방식이 분명하게 떠오르지 않아서 선택했습니다.

Livenow14 replied:
넵 그렇다면, 이대로 진행해주시면 될 것 같아요 :) 

다만 그럼에도 현재요구사항에서 라이브러리는 무겁다 생각하는 편이에요. 

다음에는 조금더 고려해보셔도 좋을 것 같아요. 기술이 모든 상황을 커버해주진 않거든요.

kdkdhoho replied:
네, 저도 동의합니다.
해당 라이브러리가 믿고 쓸만한지? 정말 필요로 한지? 여러 기준들로 따져보고 선택해야 한다고 생각해요!

Livenow14 commented on src/main/java/subway/domain/SubwayMap.java:
```diff
@@ -0,0 +1,57 @@
+        Map<Long, Line> linesMap = lines.getLines();
+        Line targetLine = linesMap.get(line.getId());
+
+        if (targetLine.isInitState()) {
+            targetLine.initStations(upStation, downStation);
+            return;
+        }
+
+        if (targetLine.hasStation(upStation)) {
+            targetLine.addLast(downStation);
+            return;
+        }
+
+        targetLine.addBeforeAt(upStation, downStation);
```
조건에 따라 line의 station 저장 로직이 달라지는 것 같은데, 
이경우 객체로 표현해보는건 어떨까요 ?

키워드: 전략패턴

Livenow14 reviewed (changes requested):
안녕하세요 도기 :) 구현잘해주셨어요. 몇가지 코멘트 남겼으니 확인부탁드려요.

MR 코멘트에 남겨주신 답변은, 코치분들과 논의 후 코드에 대판 피드백만 남기기로 해서 추가적으로 남기진 않았어요 :) 

구현되어 있지 않아, 현재레벨에서 구현까지 완료해보셨으면 좋을 것 같아요 :) 

추가적으로 피드백 남긴다면,
1. 테스트 가 깨지고 있어요. 테스트를 먼저 통과하게 작성해주시면 좋을것 같아요 :) 
2. 이번 미션에 비즈니스 규칙이 많은데 테스트에 해당 규칙을 녹여주셨으면 좋겠어요!! 
-> 이렇게 작성된다면, 요구사항을 빠르고 반복적으로 검증할 수 있어요.
```java
노선에 등록되는 역의 위치는 자유롭게 지정할 수 있어야 합니다.
A-B-C 역이 등록되어 있는 노선에 D역을 등록할 경우 C역 다음에만 등록할 수 있는게 아니라
A역과 B역 사이에도 등록이 가능하고, A역 앞에도 등록을 할 수 있어야 합니다.
노선에 역이 등록될 때 거리 정보도 함께 포함되어야 합니다.
```

Livenow14 reviewed (changes requested):
안녕하세요 도기 :) 구현잘해주셨어요. 몇가지 코멘트 남겼으니 확인부탁드려요.

MR 코멘트에 남겨주신 답변은, 코치분들과 논의 후 코드에 대판 피드백만 남기기로 해서 추가적으로 남기진 않았어요 :) 

구현되어 있지 않아, 현재레벨에서 구현까지 완료해보셨으면 좋을 것 같아요 :) 

추가적으로 피드백 남긴다면,
1. 테스트 가 깨지고 있어요. 테스트를 먼저 통과하게 작성해주시면 좋을것 같아요 :) 
2. 이번 미션에 비즈니스 규칙이 많은데 테스트에 해당 규칙을 녹여주셨으면 좋겠어요!! 
-> 이렇게 작성된다면, 요구사항을 빠르고 반복적으로 검증할 수 있어요.
```java
노선에 등록되는 역의 위치는 자유롭게 지정할 수 있어야 합니다.
A-B-C 역이 등록되어 있는 노선에 D역을 등록할 경우 C역 다음에만 등록할 수 있는게 아니라
A역과 B역 사이에도 등록이 가능하고, A역 앞에도 등록을 할 수 있어야 합니다.
노선에 역이 등록될 때 거리 정보도 함께 포함되어야 합니다.
```

kdkdhoho committed:
docs(README): .http 적용 및 파일 참조 적용

kdkdhoho committed:
feat(Stations): 맨 앞에 역을 추가하는 기능 구현

kdkdhoho committed:
refactor(LineNotFoundException): 에러 메시지 추가

kdkdhoho committed:
docs(README): 도메인 객체 설계 수정

kdkdhoho committed:
feat(Distance): Distance 객체 구현

- 객체 생성
- 음수 검증
- 거리 더하기
- 거리 빼기

kdkdhoho committed:
test(StationTest): 객체 동등성 테스트 추가

kdkdhoho committed:
docs(README): 도메인 객체 설계 수정

kdkdhoho committed:
test(Line): 객체 동등성 테스트 코드 추가

kdkdhoho committed:
feat(Section): Section 구현

kdkdhoho committed:
remove(SubwayMap): JGraphT 라이브러리 사용하여 그래프 자료구조 적용한 객체 제거

kdkdhoho committed:
feat(Section): 방향을 뒤집은 새로운 객체 생성 기능 구현

kdkdhoho committed:
docs(README): 도메인 객체 설계 Distance 추가

kdkdhoho committed:
feat(Section): 검증에 필요한 여러 기능 구현

kdkdhoho committed:
feat(Sections): 여러 Section을 가지고 비즈니스 규칙을 검증하는 객체 구현

kdkdhoho committed:
test(StationDaoTest): StationDao 테스트 추가

kdkdhoho committed:
test(LineDaoTest): StationDao 테스트 추가

kdkdhoho committed:
test: Dao 전체 테스트 시 실패하는 문제 해결

kdkdhoho committed:
feat(SectionDao): SectionDao 구현

kdkdhoho committed:
feat(SectionRepository): Section 도메인 영속을 추상화하는 객체 구현

kdkdhoho committed:
feat(SectionEntity): SectionEntity 구현

kdkdhoho committed:
chore: 프로덕션 DB와 테스트용 DB를 다르게 설정

kdkdhoho committed:
chore: 노선에 역 1개 추가하는 URI 수정

kdkdhoho committed:
feat: 빈 노선에 역 2개 추가하는 기능 구현

kdkdhoho committed:
refactor: 스켈레톤 코드 복구

kdkdhoho committed:
remove: 불필요한 객체 삭제

kdkdhoho committed:
feat(GlobalExceptionHandler): 모든 예외를 처리해주는 객체 생성

kdkdhoho committed:
docs: 오탈자 수정 및 구현 완료한 기능 체크

kdkdhoho committed:
feat(Sections): 두 Sections 객체에서 중복되는 Section을 제거한 결과 반환 기능 구현

kdkdhoho committed:
feat(SectionRepository): 여러 Section들을 모두 저장하는 기능 구현

kdkdhoho committed:
feat(GlobalExceptionHandler): IllegalAddSection 핸들러 메서드 구현

kdkdhoho committed:
feat: 역을 중간에 1개 추가하는 기능 구현

kdkdhoho committed:
test(SectionIntegrationTest): 테스트 리팩터링

kdkdhoho committed:
test(SectionIntegrationTest): 역을 추가할 때 예외 케이스 추가 작성

kdkdhoho committed:
refactor: 메서드명 수정

kdkdhoho committed:
feat(SectionDao): 라인ID와 일치하는 모든 Section 조회 기능 구현

kdkdhoho committed:
feat(SectionRepository): 라인ID와 일치하는 모든 Section 조회 기능 구현

kdkdhoho committed:
feat(SectionDao): 특정 호선 ID와 일치하는 모든 데이터 삭제 기능 구현

kdkdhoho committed:
feat(SectionRepository): Section 데이터를 갱신하는 기능 구현

kdkdhoho committed:
feat(GlobalExceptionHandler): EmptyResultDataAccessException 예외를 처리하는 기능 구현

kdkdhoho committed:
refactor: 도메인 로직 수행 후 DB에 갱신하는 로직 수정

중복된 Section을 추출하여 DB에 update 하는 방식 -> 특정 line의 모든 Section을 제거하고 도메인 로직이 수행 된 후의 Sections를 모두 저장하는 방식

kdkdhoho committed:
feat: 호선에 역을 제거하는 기능 구현

kdkdhoho committed:
feat: 노선을 조회하는 기능 구현

kdkdhoho commented:
검프, 그동안 안녕하셨나요? 1단계 제출 후 오랜 시간이 걸렸네요.
월요일까지 도메인 객체 설계에 대해 많은 고민을 하다가 조금 늦게 구현을 시작했습니다.
이번 리뷰도 감사합니다🙇‍♂️

# 질문

## Service 테스트를 어떻게 잘할 수 있을까요?
Service 테스트를 진행하면서 무엇을 어떻게 테스트할까 고민을 해보았어요.
그 결과 현재 비즈니스 로직이 "잘 수행"되는지 "행위" 테스트만 하는 쪽으로 결정했습니다.
그런데 이게 과연 의미가 있을까는 생각이 들었어요.

테스트가 프로덕션 코드에 지나치게 의존하고 있으며, 내가 짠 로직이 정상적으로 실행되는지 테스트하는 것으로 밖에 생각되지 않습니다.
동시에 비즈니스 규칙을 검증하는 작은 단위의 단위 테스트와, API 요청 및 응답, 그리고 DB에 정상적으로 저장이 되는지 확인하는 통합 테스트만 잘 작성하면 되는거 아닌가? 하는 생각이 들었습니다.

Service는 무엇을 테스트 해야하고, 어떻게 테스트를 잘 할 수 있을까요?

## SectionRepository에서 Sections를 반환해도 되나요?
현재 `Section`에 대한 도메인과 엔티티가 분리된 상황입니다. 따라서 `SectionRepository`가 존재하는데요.
동시에 `Sections` 라는 `Section`의 일급 컬렉션 도메인 객체도 존재합니다.

`SectionRepository`에서 `Sections`를 반환해도 괜찮은 지가 궁금합니다.

kdkdhoho committed:
feat(TopDownStationArranger): 상행에서 하행으로 역들을 정렬해주는 도메인 객체 구현

kdkdhoho committed:
refactor: 커스텀 예외 리팩터링 및 예외 처리 수정

kdkdhoho committed:
refactor: 도메인 객체 전반적인 리팩터링

kdkdhoho committed:
refactor: 전반적 리팩터링

- 사용하지 않는 예외 삭제
- DTO 객체에 final 키워드 추가
- 패키지명 변경

kdkdhoho committed:
refactor(Section): 전반적인 리팩터링

kdkdhoho committed:
remove(SectionServiceTest): 서비스 테스트 삭제

kdkdhoho committed:
refactor(SectionService): 중복 코드 메서드로 추출

Livenow14 commented on src/test/java/subway/domain/SectionsTest.java:
```diff
@@ -0,0 +1,290 @@
+    @Test
+    void 중간의_역이_제거되면_거리가_재배정된다() {
+        // given
+        sections.add(_2호선_신림_봉천_거리5);
+        sections.add(_2호선_봉천_낙성대_거리5);
```
중복된다면, 따로 메서드나, 셋업 메서드로 설정해도 좋겠네요 :)

kdkdhoho replied:
다음 단계에서 적용하겠습니다!

Livenow14 commented on src/main/java/subway/domain/SectionEntity.java:
```diff
@@ -0,0 +1,57 @@
+        return distance;
+    }
+
+    @Override
+    public boolean equals(final Object o) {
+        if (this == o) return true;
+        if (o == null || getClass() != o.getClass()) return false;
+        final SectionEntity that = (SectionEntity) o;
+        return id.equals(that.id);
+    }
+
+    @Override
+    public int hashCode() {
+        return Objects.hash(id);
+    }
+}
```
💯

kdkdhoho replied:
엔티티 객체는 id로만 동등성 비교하면 되나요?

사실 확신 없이 스스로 판단했을 때 id로만 비교해도 괜찮다고 생각했기에 이런 결정을 했거든요!

Livenow14 replied:
넵 맞습니다, 엔티티는 에플리케이션 내에서 식별성과 변경가능성을 가지는 특수한 객체이기에, 
식별할 수 있는 값만을 비교해야합니다. 잘 구현해주셨어요.

Livenow14 commented on src/main/java/subway/domain/Sections.java:
```diff
@@ -0,0 +1,173 @@
+    private final List<Section> sections;
+    private final TopDownStationArranger topDownStationArranger = new TopDownStationArranger();
+
+    public Sections(final List<Section> sections) {
+        this.sections = new ArrayList<>(sections);
+    }
```
💯

Livenow14 commented on src/main/java/subway/domain/Sections.java:
```diff
@@ -0,0 +1,173 @@
+    private static final int ONLY_ONE_SECTION_INDEX = 0;
+
+    private final List<Section> sections;
+    private final TopDownStationArranger topDownStationArranger = new TopDownStationArranger();
+
```
해당 객체가 존재하는게 단순 Sections의 책임을 대신 수행해주기 위함일까요?

kdkdhoho replied:
처음엔 `Sections`의 코드가 200~300줄 가까이 되었습니다. 따라서 책임(역 정렬)을 분리해야겠다는 생각이 들었어요.

추가로 클라이언트의 요청에 따라 역의 순서를 **하행에서 상행** 순으로 정렬해야하는 상황이 올 수 있다고 판단했어요.
이때는 `StationArranger` 라는 인터페이스를 두고, 그 아래에 정렬 기준을 다르게 갖는 `TopDown~`과 `BottomUp~`를 두어 사용자의 요청에 좀 더 유연하게 대응할 수 있다는 이유도 있습니다.

Livenow14 replied:
해당 요구사항이 정말 요구된 요구사항일까요 아니면, 추측에 근거된 요구사항일까요?

kdkdhoho replied:
현재까진 요구되지 않았으며, 추측에 근거한 요구사항입니다.

Livenow14 replied:
정말로 필요해지는 순간에 적절히 사용해보는건 어떨까요? 
그때 작성되어도 충분히 늦지 않고, 잘 구현해주실 것 같아서요 😄

kdkdhoho replied:
그러면 다시 처음으로 돌아가서, `TopDownStationArranger `가 `Sections`의 책임을 나눠 가질 필요는 없다고 보시나요?

Livenow14 commented on src/main/java/subway/dto/AddTwoSectionRequest.java:
```diff
@@ -0,0 +1,42 @@
+    private final Integer upStationDistance;
+    private final Integer downStationDistance;
+
+    private AddTwoSectionRequest() {
+        this(null, null, null, null, null);
+    }
```
gradle 설정에서 Build and run using을 gradle로 사용하는건 어떨까요?

Livenow14 replied:
<img width="979" alt="image" src="https://github.com/woowacourse/jwp-subway-path/assets/48986787/ccce03d6-b7a0-40c6-8ac3-ddacf8819c33">

kdkdhoho replied:
네! 현재 저는 말씀하신 설정으로 하고 있습니다.
하지만 다른 개발자가 다른 설정에서 동일한 실행을 했을 때 문제가 없길 희망하고 작성했습니다.

Livenow14 replied:
다른 개발자를 고려하기 위한 더 좋은 방법은 무엇이 있을까요?

저는 docs에서 실행하기 위해 필요한 설정들을 적어둘 것 같아요 :)

kdkdhoho replied:
해당 방법도 좋은 방법인 것 같아요!
하지만, 애초에 문제가 생기지 않도록 하는 것이 더 나은 배려라고 생각합니다. 굳이 문서화 할 필요가 없기도 하구요.
검프의 생각은 어떠신가요?

Livenow14 replied:
사용하시는 라이브러리, 프레임워크를 생각해보죠. 

해당 기술들이 사용자들의 모든 환경을 생각하고 작성되었을까요, 아니면 필요 환경을 요구할까요? 

더 나아간다면, intellij 를 사용하지 않는 eclipse 사용자들까지 사용할 수 있게 고려되어야하는걸까요? 

물론 배려의 측면을 생각하는 것은 좋은 시각이라 봅니다. 하지만 해당 배려는 정도가 없다고 생각됩니다. (어디까지 배려되어야 하는지 정의할 수 없음)

Livenow14 commented on src/main/java/subway/exception/IllegalRemoveSectionException.java:
```diff
@@ -0,0 +1,14 @@
+
+public class IllegalRemoveSectionException extends RuntimeException {
+
+    static final String MESSAGE = "구간 삭제 과정 중 예외가 발생했습니다.";
```
해당 예외를 사용하는 클라이언트에서 메세지를 주입해줄순 없을까요?

kdkdhoho replied:
해당 메시지는 `ExceptionHandler`에 의해 처리될 때 클라이언트에게 반환될 때 보여주기 위한 메시지입니다.
동시에 도메인 레벨에서는 위 예외를 발생할 때, 어떤 이유인지 간단한 설명을 주입하여 터트리는데요. 이는 개발자가 예외가 발생한 원인을 쉽게 파악하기 위해 나눠보았습니다.
서비스를 사용하는 유저 입장에서는 예외가 어떤 이유에서 터졌는지 자세히 알 필요는 없으니까요.

Livenow14 replied:
제가 말씀드린 클라이언트의 정의는 "해당 객체를 의존하여 사용하는 객체" 입니다. 

구현 코드 내에서 IllegalRemoveSectionException 와 같은 exception이 많아진다면, 매번 해당 클래스 내부를 들어와야 그 의미를 알 수 있지 않을까요 ?! 

++ 예외 메세지를 변경해달라는 요구사항이 들어왔을 때, 객체 이름 + 내부 Messsage 둘다 수정되어야 하는걸까요?

kdkdhoho replied:
`Sections` 에서 위 예외를 사용하고 있는데요. 예외를 발생할 때 메시지를 주입하고 있습니다.

++ 어떤 상황에서 어떠한 이유로 변경 요청이 들어왔냐에 따라 달라진다고 생각해요!

Livenow14 commented on src/test/java/subway/domain/SectionsTest.java:
```diff
@@ -0,0 +1,290 @@
+
+class SectionsTest {
+
+    private Sections sections;
```
해당 객체를 공유함으로써 생기는 문제는 없을까요?

kdkdhoho replied:
객체의 상태가 변경되면, 다른 테스트에도 영향을 끼치는 문제가 발생할 것 같습니다.
이를 해결하고자 `@BeforeEach` 를 통해 매번 테스트가 실행되기 전 새로운 객체로 할당 해주었습니다 !

Livenow14 reviewed (approved):
안녕하세요 도기! 구현잘해주셨어요 :) 

테스트를 잘 작성해주셔서 각 메서드의 역할을 알기쉬웠어요.
몇가지 간단한 피드백 남겼으니, 다음 미션 진행하면서 반영해주시면 좋을 것 같아요.

++ given , expect 패턴을 사용한 것은 본인만의 방식일까요?!?! 신선해서 질문드립니다 ㅎㅎ

kdkdhoho commented:
> ++ given , expect 패턴을 사용한 것은 본인만의 방식일까요?!?! 신선해서 질문드립니다 ㅎㅎ

예전 페어에게 배운 패턴입니다! 꽤 괜찮다고 판단해서 아직까지 사용하고 있어요 😀
